### PR TITLE
fix overused judgement when deal with allocate and proportion

### PR DESCRIPTION
**Pull request by Thor-wl**
fix: https://github.com/volcano-sh/volcano/issues/1425
Signed-off-by: Thor-wl <13164644535@163.com>

### CODE REVIEW

File: pkg/scheduler/actions/allocate/allocate.go [hunk 1]
```diff
@@ -137,9 +137,13 @@ func (alloc *Action) Execute(ssn *framework.Session) {
...
 		// and the PriorityQueue have no ability to update priority for a special queue.
 		var queue *api.QueueInfo
+		overusedMap := make(map[string]float64)
 		for queueID := range queueInNamespace {
 			currentQueue := ssn.Queues[queueID]
-			if ssn.Overused(currentQueue) {
+			overused, overusedMap := ssn.Overused(currentQueue)
+			_, cpuOverused := overusedMap["MilliCPU"]
+			_, memoryOverused := overusedMap["Memory"]
+			if overused && (cpuOverused || memoryOverused) {
```
Comment (wpeng102) [inline pkg/scheduler/actions/allocate/allocate.go:146]: already used `overused` why need check `cpuOverused` or  `memoryOverused`?

Comment (Thor-wl) [reply pkg/scheduler/actions/allocate/allocate.go:146]: > already used `overused` why need check `cpuOverused` or `memoryOverused`?

If basic resources are overused such as CPU and memory, no later jobs in the queue can be allocated resource any more. If scalar resource is overused such as GPU, we should judge what the later jobs in the queue needs. If it only request CPU, it's reasonable to allocate resource for it.

File: pkg/scheduler/actions/allocate/allocate.go [hunk 2]
```diff
@@ -275,3 +287,15 @@ func (alloc *Action) Execute(ssn *framework.Session) {
 }
 
 func (alloc *Action) UnInitialize() {}
+
+// isScalarResourceOverused returns true if any dimension resource is overused when comparing job and overusedMap
+func isScalarResourceOverused(job *api.JobInfo, overusedMap map[string]float64) bool {
```
Comment (william-wang) [inline pkg/scheduler/actions/allocate/allocate.go:292]: The scalar resource overuse function should be implemented in session.

**Comment by huone1** [comment]
I think it is better as the following：
1. queue resource overuse return true when all resource overuse， otherwise return false
2. add job overuse function and compare job resource list with queue resource list base on job resource list
3.  cpu 、momery must be in the comparison list
@Thor-wl

**Comment by Thor-wl** [comment]
> I think it is better as the following：
> 
> 1. queue resource overuse return true when all resource overuse， otherwise return false
> 2. add job overuse function and compare job resource list with queue resource list base on job resource list
> 3. cpu 、momery must be in the comparison list
> 4. the new function is created in the resource_info.go or session.go
>    @Thor-wl

I don't think it's reasonable to consider queue resource overused only when all dimension resource overused. It's not helpful to judge overused at some scenarios.

File: pkg/scheduler/actions/allocate/allocate.go [hunk 3]
```diff
@@ -170,6 +170,14 @@ func (alloc *Action) Execute(ssn *framework.Session) {
 		}
 
 		job := jobs.Pop().(*api.JobInfo)
+		if ssn.OverusedAfterJobEnqueue(job) {
+			jobs.Push(job)
+			klog.V(3).Infof("Queue <%s> is overused when considering job <%s>, ignore it.", queue.Name, job.Name)
+			delete(queueInNamespace, queue.UID)
+			namespaces.Push(namespace)
+			continue
```
Comment (huone1) [inline pkg/scheduler/actions/allocate/allocate.go:178]: Don't delete queue from queueInNamespace bacause there are other jobs in queue;
the job in status overuse shouldn't be re-pushed.

File: pkg/scheduler/plugins/proportion/proportion.go [hunk 4]
```diff
@@ -235,18 +235,36 @@ func (pp *proportionPlugin) OnSessionOpen(ssn *framework.Session) {
...
+		} else {
+			isAllDimensionOverUsed = false
 		}
+		metrics.UpdateQueueOverused(attr.name, isOverUsed)
+
+		return isOverUsed, isAllDimensionOverUsed
+	})
 
-		return overused
+	ssn.AddOverusedAfterJobEnqueueFn(pp.Name(), func(taskInfo *api.TaskInfo, queueID api.QueueID) bool {
```
Comment (william-wang) [inline pkg/scheduler/plugins/proportion/proportion.go:263]: Change "AddOverusedAfterJobEnqueueFn" to "AddOverusedForTask"

File: pkg/scheduler/framework/session_plugins.go [hunk 5]
```diff
@@ -241,19 +246,36 @@ func (ssn *Session) Preemptable(preemptor *api.TaskInfo, preemptees []*api.TaskI
...
+				return true, isAllDimensionOverUsed
 			}
 		}
 	}
 
+	return false, false
+}
+
+// OverusedAfterJobEnqueue invoke overused function of the plugins
+func (ssn *Session) OverusedAfterJobEnqueue(taskInfo *api.TaskInfo, queueID api.QueueID) bool {
```
Comment (william-wang) [inline pkg/scheduler/framework/session_plugins.go:267]: Change "OverusedAfterJobEnqueue" to "OverusedForTask"

File: pkg/scheduler/plugins/proportion/proportion.go [hunk 6]
```diff
@@ -235,18 +235,36 @@ func (pp *proportionPlugin) OnSessionOpen(ssn *framework.Session) {
 		return victims, util.Permit
 	})
 
-	ssn.AddOverusedFn(pp.Name(), func(obj interface{}) bool {
-		queue := obj.(*api.QueueInfo)
+	ssn.AddOverusedFn(pp.Name(), func(queue *api.QueueInfo) (bool, bool) {
```
Comment (william-wang) [inline pkg/scheduler/plugins/proportion/proportion.go:238]: Whether this function can be simplied as below.
{
    ...
    isAllDimensionOverUsed := attr.deserved.LessEqual(attr.allocated, api.Zero)
    isOverUsed := !attr.allocated.LessEqual(attr.deserved, api.Zero)
    metrics.UpdateQueueOverused(attr.name, isOverUsed)
   ...
return isOverUsed, isAllDimensionOverUsed
}

Comment (Thor-wl) [reply pkg/scheduler/plugins/proportion/proportion.go:238]: Good idea. Let me optimize it

File: pkg/scheduler/plugins/proportion/proportion.go [hunk 7]
```diff
@@ -235,18 +235,22 @@ func (pp *proportionPlugin) OnSessionOpen(ssn *framework.Session) {
...
-		}
+		return isOverUsed, isAllDimensionOverUsed
+	})
+
+	ssn.AddOverusedForTask(pp.Name(), func(taskInfo *api.TaskInfo, queueID api.QueueID) bool {
+		podRequest := taskInfo.Resreq.Clone()
+		queueAllocated := pp.queueOpts[queueID].allocated.Clone()
 
-		return overused
+		return !podRequest.AddBaseLeft(queueAllocated).LessEqual(pp.queueOpts[queueID].deserved, api.Zero)
```
Comment (william-wang) [inline pkg/scheduler/plugins/proportion/proportion.go:253]: Why need to add "AddBaseLeft" func here?

File: pkg/scheduler/api/types.go [hunk 8]
```diff
@@ -168,3 +168,9 @@ type ReservedNodesFn func()
 
 // VictimTasksFn is the func declaration used to select victim tasks
 type VictimTasksFn func() []*TaskInfo
+
+// QueueOverusedFn is the func declaration used to judge whether queue is overused.If overused, whether it is all dimensions overused.
+type QueueOverusedFn func(*QueueInfo) (bool, bool)
+
+// QueueOverusedAfterJobEnqueueFn is the func declaration used to judge whether queue is overused after pod is allocated resource.
+type QueueOverusedAfterJobEnqueueFn func(*TaskInfo, QueueID) bool
```
Comment (william-wang) [inline pkg/scheduler/api/types.go:176]: change "QueueOverusedAfterJobEnqueueFn" to "QueueOverusedForTaskFn"

File: pkg/scheduler/framework/session.go [hunk 9]
```diff
@@ -57,31 +57,32 @@ type Session struct {
...
+	predicateFns               map[string]api.PredicateFn
+	bestNodeFns                map[string]api.BestNodeFn
+	nodeOrderFns               map[string]api.NodeOrderFn
+	batchNodeOrderFns          map[string]api.BatchNodeOrderFn
+	nodeMapFns                 map[string]api.NodeMapFn
+	nodeReduceFns              map[string]api.NodeReduceFn
+	preemptableFns             map[string]api.EvictableFn
+	reclaimableFns             map[string]api.EvictableFn
+	overusedFns                map[string]api.QueueOverusedFn
+	overusedAfterJobEnqueueFns map[string]api.QueueOverusedAfterJobEnqueueFn
```
Comment (william-wang) [inline pkg/scheduler/framework/session.go:76]: s/overusedAfterJobEnqueueFns/overusedForTaskFns/g

File: pkg/scheduler/plugins/proportion/proportion.go (hunk 7 above)
Comment (huone1) [reply pkg/scheduler/plugins/proportion/proportion.go:253]: if the GPU in queueAllocated is overuse, the GPU  must be overuse after podRequest + queueAllocated. So it is add besed on the  resourcelist of podRequest

**Comment by k82cn** [comment]
let's review the API firstly before merge it :)

**Comment by Thor-wl** [comment]
> let's review the API firstly before merge it :)

api modification: https://github.com/volcano-sh/volcano/pull/1676

File: pkg/scheduler/actions/reclaim/reclaim.go [hunk 10]
```diff
@@ -91,7 +91,8 @@ func (ra *Action) Execute(ssn *framework.Session) {
 		var task *api.TaskInfo
 
 		queue := queues.Pop().(*api.QueueInfo)
-		if ssn.Overused(queue) {
+		overused := ssn.Overused(queue)
```
Comment (william-wang) [inline pkg/scheduler/actions/reclaim/reclaim.go:94]: The overuse in reclaim might face the similar issue with alloate action.  For example
there are queue1 and queue2 in cluster.  The existing jobs in queue1 occurpy all cpu resource. The existing jobs in queue2 occupy all gpu resource. And then we submit a cpu job in queue2.  Is the cpu job in queue2 supposed to reclaim resource from queue1?

File: pkg/scheduler/plugins/proportion/proportion.go (hunk 7 above)
Comment (william-wang) [reply pkg/scheduler/plugins/proportion/proportion.go:262]: Got it. The logic is a little trick.
AddbaseLeft function
**input:** 
r  cpu 1
rr cpu1 gpu1
**output:**
r  cpu: 2

Add function
**input:** 
r: cpu 1
rr: cpu1 gpu1
**output:** 
r  cpu: 2, gpu: 1

File: pkg/scheduler/actions/reclaim/reclaim.go [hunk 11]
```diff
@@ -111,6 +111,15 @@ func (ra *Action) Execute(ssn *framework.Session) {
 			task = tasks.Pop().(*api.TaskInfo)
 		}
 
+		// some resource dimensions task requests are overused in queue
```
Comment (william-wang) [inline pkg/scheduler/actions/reclaim/reclaim.go:114]: // some resource dimensions task requests are overused in queue ->
// Check whether the queue is overused on dimension that the task requested.

File: pkg/scheduler/api/resource_info.go [hunk 12]
```diff
@@ -567,3 +577,29 @@ func ParseResourceList(m map[string]string) (v1.ResourceList, error) {
...
+func GetMinResource() float64 {
+	return minResource
+}
+
+// ResourceNames struct defines resource name collection
+type ResourceNames struct {
+	Names []v1.ResourceName
+}
+
+// Subset judges whether dimensions in r is subset of that of rr
```
Comment (william-wang) [inline pkg/scheduler/api/resource_info.go:590]: // Subset judges whether dimensions in r is subset of that of rr
->
// Subset judges whether r is part of rr

File: pkg/scheduler/api/resource_info.go [hunk 13]
```diff
@@ -567,3 +577,31 @@ func ParseResourceList(m map[string]string) (v1.ResourceList, error) {
...
+
+func GetMinResource() float64 {
+	return minResource
+}
+
+// ResourceNameList struct defines resource name collection
+type ResourceNameList []v1.ResourceName
+
+// Contains judges whether rr is subset of r
+func (r *ResourceNameList) Contains(rr *ResourceNameList) bool {
```
Comment (k82cn) [inline pkg/scheduler/api/resource_info.go:589]: let's use reference directly.

File: pkg/scheduler/api/types.go [hunk 14]
```diff
@@ -168,3 +168,6 @@ type ReservedNodesFn func()
 
 // VictimTasksFn is the func declaration used to select victim tasks
 type VictimTasksFn func() []*TaskInfo
+
+// UnderUsedFn is the func declaration used to get under used resource list for queue
+type UnderUsedFn func(*QueueInfo) *ResourceNameList
```
Comment (k82cn) [inline pkg/scheduler/api/types.go:173]: s/UnderUsedFn/UnderusedResourceFn/

File: pkg/scheduler/api/resource_info.go [hunk 15]
```diff
@@ -114,10 +114,20 @@ func (r *Resource) String() string {
 
 // ResourceNames returns all resource types
 func (r *Resource) ResourceNames() []v1.ResourceName {
```
Comment (k82cn) [inline pkg/scheduler/api/resource_info.go:116]: return `ResourceNameList`

File: pkg/scheduler/plugins/proportion/proportion.go [hunk 16]
```diff
@@ -249,6 +250,28 @@ func (pp *proportionPlugin) OnSessionOpen(ssn *framework.Session) {
 		return overused
 	})
 
+	ssn.AddUnderusedFn(pp.Name(), func(queue *api.QueueInfo) api.ResourceNameList {
```
Comment (k82cn) [inline pkg/scheduler/plugins/proportion/proportion.go:253]: also rename to UnderusedResourceFn

File: pkg/scheduler/framework/session_plugins.go [hunk 17]
```diff
@@ -103,6 +103,11 @@ func (ssn *Session) AddOverusedFn(name string, fn api.ValidateFn) {
 	ssn.overusedFns[name] = fn
 }
 
+// AddUnderusedFn add underused function
+func (ssn *Session) AddUnderusedFn(name string, fn api.UnderUsedResourceFn) {
```
Comment (k82cn) [inline pkg/scheduler/framework/session_plugins.go:107]: align with UnderusedResources

**Comment by k82cn** [comment]
/hold cancel

LGTM overall :)

File: pkg/scheduler/api/resource_info.go [hunk 18]
```diff
@@ -567,3 +577,32 @@ func ParseResourceList(m map[string]string) (v1.ResourceList, error) {
...
+			return false
+		}
+	}
+	return true
+}
+
+// Append adds new resource name to r
+func (r ResourceNameList) Append(rName v1.ResourceName) {
+	r = append(r, rName)
+}
```
Comment (william-wang) [inline pkg/scheduler/api/resource_info.go:608]: The Append is not able to modify the ResourceNameList.

File: pkg/scheduler/plugins/proportion/proportion.go [hunk 19]
```diff
@@ -249,6 +250,28 @@ func (pp *proportionPlugin) OnSessionOpen(ssn *framework.Session) {
...
+		}
+		if underUsedResource.Memory >= api.GetMinResource() {
+			underUsedResName.Append(v1.ResourceMemory)
+		}
+		for rName := range underUsedResource.ScalarResources {
+			underUsedResName.Append(rName)
+		}
+		klog.V(3).Infof("Queue <%v>: deserved <%v>, allocated <%v>, share <%v>, underUsedResName %v",
+			queue.Name, attr.deserved, attr.allocated, attr.share, underUsedResName)
+		// metrics.UpdateQueueOverused(attr.name, isOverUsed)
```
Comment (william-wang) [inline pkg/scheduler/plugins/proportion/proportion.go:269]: remove the commented line.

File: pkg/scheduler/plugins/proportion/proportion.go [hunk 20]
```diff
@@ -249,6 +250,28 @@ func (pp *proportionPlugin) OnSessionOpen(ssn *framework.Session) {
 		return overused
 	})
 
+	ssn.AddUnderusedFn(pp.Name(), func(queue *api.QueueInfo) api.ResourceNameList {
+		underUsedResName := api.ResourceNameList{}
```
Comment (william-wang) [inline pkg/scheduler/plugins/proportion/proportion.go:254]: s/underUsedResName/underUsedResNames/g

**Pull request merged**